src/utils/process.ts

```diff
@@ -0,0 +1,101 @@
+import { spawn } from 'child_process'
+import { Readable } from 'stream'
+
+interface ProcessResult {
+  code: number | null
+  signal: string | null
+  stdout: string
+  stderr: string
+}
+
+interface ProcessOptions {
+  detached?: boolean
+  env?: NodeJS.ProcessEnv
+  maxSize?: number
+  stdOutMaxSize?: number
+  stdErrMaxSize?: number
+  timeout?: number
+  closeStdin?: boolean
+}
+
+interface ExitResult {
+  code: number | null
+  signal: string | null
+}
+
+export async function runProcessAsync(
+  command: string,
+  args: string[],
+  options: ProcessOptions = {}
+): Promise<ProcessResult> {
+  const cmd = spawn(command, args, {
+    env: options.env,
+    detached: options.detached
+  })
+
+  const maxSize = options.maxSize || 10 * 1024 * 1024
+  const stdOutMaxSize = options.stdOutMaxSize || maxSize
+  const stdErrMaxSize = options.stdErrMaxSize || maxSize
+
+  const promise = new Promise<ProcessResult>((resolve, reject) => {
+    if (options.timeout) {
+      setTimeout(() => {
+        reject(new Error('Timeout'))
+        cmd.kill() // Does not terminate the node process in the shell
+      }, options.timeout)
+    }
+
+    const stdoutPromise = readAllAsync(cmd.stdout, stdOutMaxSize)
+    const stderrPromise = readAllAsync(cmd.stderr, stdErrMaxSize)
+
+    // Close stdin
+    if (options.closeStdin) {
+      cmd.stdin.end()
+    }
+
+    const exitPromise = new Promise<ExitResult>(resolve => {
+      cmd.on('exit', (code, signal) => {
+        resolve({ code, signal })
+      })
+    })
+
+    Promise.all([exitPromise, stdoutPromise, stderrPromise])
+      .then(results => {
+        resolve({
+          code: results[0]['code'],
+          signal: results[0]['signal'],
+          stdout: results[1].toString('utf8'),
+          stderr: results[2].toString('utf8')
+        })
+      })
+      .catch(reject)
+  })
+
+  return promise.then((result: ProcessResult) => result)
+}
+
+async function readAllAsync(fd: Readable, maxSize: number): Promise<Buffer | string> {
+  return new Promise((resolve, reject) => {
+    const data: (Buffer | string)[] = []
+    let dataLength = 0
+
+    fd.on('data', (chunk: Buffer | string) => {
+      dataLength += chunk.length
+
+      if (dataLength > maxSize) {
+        reject(new Error(`Data size larger than maxsize: ${dataLength} > ${maxSize}`))
+        fd.destroy()
+      }
+
+      data.push(chunk)
+    })
```

Comment:
Small nitpick. 
We don't have a standard for this but we should probably not reject in the `on('data')` and instead destroy with an error which is then emitted. 
```suggestion
    fd.on('data', (chunk: Buffer | string) => {
      dataLength += chunk.length

      if (dataLength > maxSize) {
        fd.destroy(new Error(`Data size larger than maxsize: ${dataLength} > ${maxSize}`))
      }
      data.push(chunk)
    })
    fd.on('error', reject)
```

Reply:
I generally tried to not update any code except to fix type errors since this package is a critical part of our infrastructure. I don't have a lot of experience with streams but destroying with an error emits an `'error'` event and relinquishes all resources of stream but wouldn't that mean that the promise never rejects or even resolves since it also emits a `'close'` event but not necessarily an `'end'`? I tried to out your suggestion in the test and the test exceeded the default timeout of 5 seconds because it never resolves.

The original code actually destroys and then rejects so I'll reinstate that change.

Reply:
Alright. Lets just stick with the previous change